File: src/bun/bun_serve/src/routes/health.ts
import { databaseService } from '../services/database.ts';
import { cacheService } from '../services/cache.ts';
import type { HealthStatus } from '../types.ts';

export async function healthHandler(request: Request): Promise<Response> {
  try {
    const dbHealthy = await databaseService.healthCheck();
    const cacheHealthy = await cacheService.ping();

    const health: HealthStatus = {
      status: dbHealthy && cacheHealthy ? 'healthy' : 'unhealthy',
      version: '1.0.0',
      timestamp: new Date().toISOString(),
      database: dbHealthy ? 'healthy' : 'unhealthy',
      cache: cacheHealthy ? 'healthy' : 'unhealthy',
    };

    const statusCode = dbHealthy && cacheHealthy ? 200 : 503;

    return new Response(JSON.stringify(health), {
      status: statusCode,
      headers: {
        'Content-Type': 'application/json',
        'Cache-Control': 'no-cache',
      },
    });
  } catch (error) {
    return new Response(JSON.stringify({
      status: 'error',
      version: '1.0.0',
      timestamp: new Date().toISOString(),
      database: 'unhealthy',
      cache: 'unhealthy',
      error: 'Internal server error',
    }), {
      status: 500,
      headers: {
        'Content-Type': 'application/json',
      },
    });
  }
}

export async function healthzHandler(request: Request): Promise<Response> {
  return new Response(JSON.stringify({ status: 'ok' }), {
    status: 200,
    headers: {
      'Content-Type': 'application/json',
    },
  });
}

File: src/go/chi/handlers/health.go
package handlers

import (
	"database/sql"
	"encoding/json"
	"net/http"
	"time"

	"chi/services"
)

func HealthHandler(db *sql.DB, cacheService *services.CacheService) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")

		dbStatus := "disconnected"
		if db != nil {
			if err := db.Ping(); err == nil {
				dbStatus = "connected"
			}
		}

		cacheStatus := "disconnected"
		if cacheService != nil {
			if err := cacheService.Ping(); err == nil {
				cacheStatus = "connected"
			}
		}

		status := "healthy"
		if dbStatus == "disconnected" || cacheStatus == "disconnected" {
			status = "unhealthy"
		}

		json.NewEncoder(w).Encode(map[string]interface{}{
			"status":    status,
			"version":   "1.0.0",
			"timestamp": time.Now().UTC().Format(time.RFC3339),
			"database":  dbStatus,
			"cache":     cacheStatus,
		})
	}
}

func HealthzHandler(w http.ResponseWriter, r *http.Request) {
	w.WriteHeader(http.StatusOK)
	w.Write([]byte("OK"))
}
